File: fixtures/coding/maintainability-mini-service-go/repository.go
package miniservice

import "strings"

// CustomerRepository is a small repository with deterministic IDs for testability.
type CustomerRepository struct {
	customers []Customer
	nextID    int
}

func NewCustomerRepository() *CustomerRepository {
	return &CustomerRepository{nextID: 1}
}

func (r *CustomerRepository) ListCustomers() []Customer {
	customers := make([]Customer, len(r.customers))
	copy(customers, r.customers)
	return customers
}

func (r *CustomerRepository) FindByEmail(email string) *Customer {
	normalized := strings.ToLower(strings.TrimSpace(email))
	for index := range r.customers {
		if r.customers[index].Email == normalized {
			customer := r.customers[index]
			return &customer
		}
	}
	return nil
}

func (r *CustomerRepository) CreateCustomer(name string, email string, plan string, tags []string) Customer {
	customer := Customer{
		ID:    "cus_" + itoa(r.nextID),
		Name:  name,
		Email: strings.ToLower(strings.TrimSpace(email)),
		Plan:  plan,
		Tags:  append([]string{}, tags...),
	}
	r.nextID++
	r.customers = append(r.customers, customer)
	return customer
}

func itoa(value int) string {
	if value == 0 {
		return "0"
	}
	digits := []byte{}
	for value > 0 {
		digits = append([]byte{byte('0' + value%10)}, digits...)
		value /= 10
	}
	return string(digits)
}

File: fixtures/coding/maintainability-mini-service-rust/src/repository.rs
use crate::models::Customer;

/// Small repository with deterministic IDs for testability.
#[derive(Clone, Debug, Default)]
pub struct CustomerRepository {
    customers: Vec<Customer>,
    next_id: usize,
}

impl CustomerRepository {
    pub fn new() -> Self {
        Self {
            customers: Vec::new(),
            next_id: 1,
        }
    }

    pub fn list_customers(&self) -> Vec<Customer> {
        self.customers.clone()
    }

    pub fn find_by_email(&self, email: &str) -> Option<Customer> {
        let normalized = email.trim().to_lowercase();
        self.customers
            .iter()
            .find(|customer| customer.email == normalized)
            .cloned()
    }

    pub fn create_customer(
        &mut self,
        name: String,
        email: String,
        plan: String,
        tags: Vec<String>,
    ) -> Customer {
        let customer = Customer {
            id: format!("cus_{}", self.next_id),
            name,
            email: email.trim().to_lowercase(),
            plan,
            tags,
        };
        self.next_id += 1;
        self.customers.push(customer.clone());
        customer
    }
}
